tests/cases/compiler/typeGuardConstructor.ts

```diff
@@ -0,0 +1,126 @@
+// Typical case
+class Foo {
+    prop1: string;
+}
+
+let foo: Foo | number;
+if (foo.constructor == Foo) {
+    foo.prop1; // string
+}
+if (foo["constructor"] == Foo) {
+    foo.prop1; // string
+}
+if (foo.constructor === Foo) {
+    foo.prop1; // string
+}
+if (foo["constructor"] === Foo) {
+    foo.prop1; // string
+}
+if (Foo == foo.constructor) {
+    foo.prop1; // string
+}
+if (Foo == foo["constructor"]) {
+    foo.prop1; // string
+}
+if (Foo === foo.constructor) {
+    foo.prop1; // string
+}
+if (Foo === foo["constructor"]) {
+    foo.prop1; // string
+}
+
+if (foo.constructor != Foo) {
+    foo.prop1; // ERROR
+}
+if (foo["constructor"] != Foo) {
+    foo.prop1; // ERROR
+}
+if (foo.constructor !== Foo) {
+    foo.prop1; // ERROR
+}
+if (foo["constructor"] !== Foo) {
+    foo.prop1; // ERROR
+}
+if (Foo != foo.constructor) {
+    foo.prop1; // ERROR
+}
+if (Foo != foo["constructor"]) {
+    foo.prop1; // ERROR
+}
+if (Foo !== foo.constructor) {
+    foo.prop1; // ERROR
+}
+if (Foo !== foo["constructor"]) {
+    foo.prop1; // ERROR
+}
+
+
+// Derived class case
+class Bar extends Foo {
+    prop2: number;
+}
+
+let bar: Bar | boolean;
+if (bar.constructor === Bar) {
+    bar.prop1; // string
+    bar.prop2; // number
+}
+if (bar.constructor === Foo) {

```

Comment:
As far as I can tell, this should always be false. Right now the type get's narrowed to `Bar` in the if, is this right? Or should it narrow just to `Foo`?

Reply:
Shouldn't it narrow to `never` inside the if since Bar | boolean` can never satisfy the condition?

Reply:
That's a good point. I didn't think about narrowing to `never`. Seems like the thing to do. I'll change this. Thanks.